File: web_app_health/server/WebHealthServer/WebHealthServer/Models/MealEntry.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using WebHealthServer.Models.Enums;

namespace WebHealthServer.Models
{
    public class MealEntry : AbstractEntity
    {
        [Required]
        public int ClientId { get; set; }

        [ForeignKey(nameof(ClientId))]
        [InverseProperty("MealEntries")]
        public Client Client { get; set; }

        [Required]
        public DateOnly EntryDate { get; set; }

        public TimeOnly? EntryTime { get; set; } = TimeOnly.FromDateTime(DateTime.UtcNow);

        [Required]
        [MaxLength(20)]
        public string MealType { get; set; } = "other"; // breakfast, lunch, dinner, snack

        [Required]
        [MaxLength(200)]
        public string FoodName { get; set; }

        [MaxLength(100)]
        public string? Brand { get; set; }

        // Количественные показатели
        [Required]
        [Range(0.1, 10000)]
        public decimal Quantity { get; set; } // Количество

        [MaxLength(10)]
        public string Unit { get; set; } = "g"; // g, ml, pcs

        // Пищевая ценность (можно вводить вручную или рассчитывать)
        [Range(0, 10000)]
        public decimal Calories { get; set; }

        [Range(0, 1000)]
        public decimal Protein { get; set; }

        [Range(0, 1000)]
        public decimal Carbohydrates { get; set; }

        [Range(0, 1000)]
        public decimal Fat { get; set; }
        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
        public DateTime? UpdatedAt { get; set; }

        [MaxLength(50)]
        public string? FatSecretFoodId { get; set; } // 🔗 Ссылка на продукт в FatSecret

        // Дополнительно
        [MaxLength(500)]
        public string? Notes { get; set; }

        // Для быстрого поиска/агрегации
        public int Year => EntryDate.Year;
        public int Month => EntryDate.Month;
        public int Day => EntryDate.Day;
    }
}

File: web_app_health/client/web-health-service/src/types/mealEntry.ts
export interface CreateMealEntryDto {
  entryDate: string;
  entryTime?: string;
  mealType: 'breakfast' | 'lunch' | 'dinner' | 'snack' | 'other';
  foodName: string;
  brand?: string;
  quantity: number;
  unit: 'g' | 'ml' | 'pcs' | 'cup' | 'tbsp' | 'tsp' | 'oz' | 'serving';
  fatSecretFoodId?: string;
  fatSecretServingId?: string; // 🔥 NEW: ID выбранной порции
  calories?: number;
  protein?: number;
  carbohydrates?: number;
  fat?: number;
  notes?: string;
}

export interface MealEntryResponseDto {
  id: number;
  entryDate: string;
  entryTime?: string;
  mealType: string;
  foodName: string;
  quantity: number;
  unit: string;
  calories: number;
  protein: number;
  carbohydrates: number;
  fat: number;
  notes?: string;
  createdAt: string;
}

export interface DailySummaryDto {
  date: string;
  totalCalories: number;
  totalProtein: number;
  totalCarbohydrates: number;
  totalFat: number;
  mealCount: number;
  meals: MealEntryResponseDto[];
}

// 🔥 NEW: Типы для порций FatSecret
export interface FatSecretServing {
  serving_id: string;
  serving_description: string;
  serving_url: string;
  metric_serving_amount: string;
  metric_serving_unit: string;
  number_of_units: string;
  measurement_description: string;
  calories: string;
  carbohydrate: string;
  protein: string;
  fat: string;
  saturated_fat?: string;
  polyunsaturated_fat?: string;
  monounsaturated_fat?: string;
  cholesterol?: string;
  sodium?: string;
  potassium?: string;
  fiber?: string;
  sugar?: string;
}

export interface FatSecretFoodDetails {
  food_id: string;
  food_name: string;
  food_type?: string;
  brand_name?: string;
  servings: FatSecretServing[];
}

export interface FoodSearchResult {
  source: 'fatsecret' | 'personal';
  fatSecretFoodId?: string;
  personalFoodId?: number;
  name: string;
  brand?: string;
  servingSize?: number; 
  caloriesPer100g: number;
  proteinPer100g: number;
  carbsPer100g: number;
  fatPer100g: number;
  defaultUnit: 'g' | 'ml' | 'pcs';
  servingInfo?: string;
  isDetailsLoaded?: boolean;
  servings?: FatSecretServing[]; // 🔥 NEW: все порции продукта
}